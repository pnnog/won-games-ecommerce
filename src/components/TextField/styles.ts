import styled, { DefaultTheme, css } from 'styled-components';
import { TextFieldProps } from '.';

export const InputWrapper = styled.div`
  ${({ theme }) => css`
    display: flex;
    align-items: center;
    gap: ${theme.spacings.xxsmall};
    padding: ${theme.spacings.xsmall};
    border-radius: 0.2rem;

    border: 0.2rem solid ${theme.colors.lightGray};
    background-color: ${theme.colors.lightGray};

    :focus-within {
      box-shadow: 0 0 0.5rem ${theme.colors.primary};
    }
  `}
`;

type IconWrapperProps = Pick<TextFieldProps, 'iconPosition'>;

export const IconWrapper = styled.figure<IconWrapperProps>`
  ${({ theme, iconPosition }) => css`
    width: 2.2rem;
    height: 2.2rem;
    line-height: 1;

    color: ${theme.colors.gray};
    order: ${iconPosition === 'left' ? '0' : '1'};
    svg {
      height: 100%;
    }
  `}
`;
export const Input = styled.input`
  ${({ theme }) => css`
    background-color: transparent;
    color: ${theme.colors.black};
    font-family: ${theme.font.family};
    font-size: ${theme.font.sizes.medium};
    border: 0;
    width: 100%;
    line-height: ${theme.font.sizes.medium};

    :focus {
      outline: 0;
    }

    &:-webkit-autofill {
      -webkit-box-shadow: 0 0 0 ${theme.spacings.small}
        ${theme.colors.lightGray} inset;
      filter: none;

      &::first-line {
        font-family: ${theme.font.family};
        font-size: ${theme.font.sizes.medium};
      }
    }
  `}
`;
export const Label = styled.label`
  ${({ theme }) => css`
    font-size: ${theme.font.sizes.small};
    color: ${theme.colors.black};
  `}
`;

export const Error = styled.p`
  ${({ theme }) => css`
    color: ${theme.colors.red};
    font-size${theme.font.sizes.xsmall}
  `}
`;

const wrapperModifiers = {
  disabled: (theme: DefaultTheme) => css`
    ${Input}, ${Label}, ${IconWrapper} {
      color: ${theme.colors.gray};
      cursor: not-allowed;

      ::placeholder {
        color: currentColor;
      }
    }
  `,

  error: (theme: DefaultTheme) => css`
    ${InputWrapper} {
      border-color: ${theme.colors.red};
    }

    ${IconWrapper}, ${Label} {
      color: ${theme.colors.red};
    }
  `,
};
type WrapperProps = Pick<TextFieldProps, 'disabled'> & { error?: boolean };
export const Wrapper = styled.div<WrapperProps>`
  ${({ theme, disabled, error }) => css`
    display: flex;
    flex-direction: column;

    ${disabled && wrapperModifiers.disabled(theme)}
    ${error && wrapperModifiers.error(theme)}
  `}
`;
