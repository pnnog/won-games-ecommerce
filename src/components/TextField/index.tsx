import { InputHTMLAttributes, useState } from 'react';
import * as S from './styles';

export type TextFieldProps = {
  onInput?: (value: string) => void;
  disabled?: boolean;
  initialValue?: string;
  label?: string;
  labelFor?: string;
  icon?: React.ReactNode;
  iconPosition?: 'left' | 'right';
  error?: string;
} & InputHTMLAttributes<HTMLInputElement>;

const TextField = ({
  onInput,
  initialValue = '',
  label,
  labelFor,
  icon,
  disabled = false,
  iconPosition = 'left',
  error = '',
  ...props
}: TextFieldProps) => {
  const [value, setValue] = useState(initialValue);

  const handleOnChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    setValue(e.currentTarget.value);

    !!onInput && onInput(value);
  };

  return (
    <S.Wrapper disabled={disabled} error={!!error}>
      {!!label && <S.Label htmlFor={labelFor}>{label}</S.Label>}
      <S.InputWrapper>
        {!!icon && (
          <S.IconWrapper role="icon" iconPosition={iconPosition}>
            {icon}
          </S.IconWrapper>
        )}
        <S.Input
          type="text"
          id={labelFor}
          value={value}
          onChange={handleOnChange}
          disabled={disabled}
          {...props}
        />
      </S.InputWrapper>

      {!!error && <S.Error>{error}</S.Error>}
    </S.Wrapper>
  );
};

export default TextField;
