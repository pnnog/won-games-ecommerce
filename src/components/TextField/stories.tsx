import { Meta, Story } from '@storybook/react';
import TextField, { TextFieldProps } from '.';
import { Email } from '@styled-icons/material-outlined';

export default {
  title: 'forms/TextField',
  component: TextField,
  args: {},

  argTypes: {
    onInput: { action: 'typing...' },
  },

  parameters: {
    viewport: {
      defaultViewport: 'mobile1',
    },

    layout: {
      default: 'fullscreen',
    },
  },
} as Meta;

export const Default: Story<TextFieldProps> = (args) => {
  return (
    <div style={{ padding: '1rem' }}>
      <TextField {...args} />
    </div>
  );
};

Default.args = {
  label: 'E-mail:',
  labelFor: 'email',
  placeholder: 'fulano@gmail.com',
  icon: <Email />,
};

export const Disabled: Story<TextFieldProps> = (args) => {
  return (
    <div style={{ padding: '1rem' }}>
      <TextField {...args} />
    </div>
  );
};

Disabled.args = {
  label: 'E-mail:',
  labelFor: 'email',
  placeholder: 'fulano@gmail.com',
  icon: <Email />,
  disabled: true,
};

export const WithError: Story<TextFieldProps> = (args) => {
  return (
    <div style={{ padding: '1rem' }}>
      <TextField {...args} />
    </div>
  );
};

WithError.args = {
  label: 'E-mail:',
  labelFor: 'email',
  placeholder: 'fulano@gmail.com',
  icon: <Email />,
  error: 'Ops.. Something went wrong',
};
