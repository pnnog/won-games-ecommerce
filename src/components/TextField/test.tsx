import { screen, waitFor } from '@testing-library/react';
import { Email } from '@styled-icons/material-outlined';

import TextField from '.';
import { renderWithTheme } from 'utils/tests/helpers';
import userEvent from '@testing-library/user-event';

describe('<TextField />', () => {
  it('should render TextField', () => {
    const { container } = renderWithTheme(
      <TextField label="textfield" labelFor="labelFor" />
    );

    const input = screen.getByRole('textbox', { name: 'textfield' });
    const label = screen.getByText('textfield');

    expect(input).toHaveAttribute('id', 'labelFor');
    expect(label).toHaveAttribute('for', 'labelFor');

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render the text on textField when its change', async () => {
    //mock da função
    const mockOnInput = jest.fn();
    renderWithTheme(
      <TextField label="textfield" labelFor="labelFor" onInput={mockOnInput} />
    );

    const input = screen.getByRole('textbox');
    const text = 'usuário inserindo dados no input';
    userEvent.type(input, text);

    await waitFor(() => {
      expect(input).toHaveValue(text);
      expect(mockOnInput).toHaveBeenCalledTimes(text.length);
    });
  });

  it('should be acessible with tab', async () => {
    renderWithTheme(<TextField label="text field" labelFor="text-field" />);

    const input = screen.getByRole('textbox', { name: /text field/i });

    const body = document.body;
    expect(body).toHaveFocus();

    userEvent.tab();

    waitFor(() => {
      expect(input).toHaveFocus();
    });
  });

  it('should render textfield with icon on left side', () => {
    renderWithTheme(<TextField icon={<Email />} />);

    const icon = screen.getByRole('icon');

    expect(icon).toHaveStyle({ width: '2.2rem' });
    expect(icon).toHaveStyle({ order: '0' });
  });

  it('should render textfield with icon on left side', () => {
    renderWithTheme(<TextField icon={<Email />} iconPosition="right" />);

    const icon = screen.getByRole('icon');

    expect(icon).toHaveStyle({ width: '2.2rem' });
    expect(icon).toHaveStyle({ order: '1' });
  });

  it('should render textfield disabled and value not to be change', async () => {
    const mockOnInput = jest.fn();

    renderWithTheme(
      <TextField
        label="text field"
        labelFor="text-field"
        onInput={mockOnInput}
        disabled
      />
    );

    const input = screen.getByRole('textbox', { name: /text field/i });

    expect(input).toBeDisabled();
    expect(input).toHaveValue('');

    userEvent.type(input, 'testando o value do input');
    await waitFor(() => expect(input).toHaveValue(''));
    expect(mockOnInput).toHaveBeenCalledTimes(0);
  });

  it('Is not acessible by tab when disabled', async () => {
    renderWithTheme(
      <TextField label="text field" labelFor="text-field" disabled />
    );

    const input = screen.getByRole('textbox', { name: /text field/i });
    const body = document.body;

    userEvent.tab();
    expect(input).not.toHaveFocus();
    expect(body).toHaveFocus();
  });

  it('should render text field when have an error', () => {
    const { container } = renderWithTheme(
      <TextField
        label="text field"
        labelFor="text-field"
        error="error text"
        icon={<Email />}
      />
    );

    const inputWrapper = screen.getByRole('textbox', {
      name: /text field/i,
    }).parentElement;
    expect(inputWrapper).toHaveStyle({ 'border-color': '#FF6347' });

    const icon = screen.getByRole('icon');
    expect(icon).toHaveStyle({ color: '#FF6347' });

    const label = screen.getByText('text field');
    expect(label).toHaveStyle({ color: '#FF6347' });

    expect(container.firstChild).toMatchSnapshot();
  });
});
