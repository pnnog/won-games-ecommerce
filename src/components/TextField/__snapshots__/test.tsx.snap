// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextField /> should render TextField 1`] = ` `;

exports[`<TextField /> should render text field when have an error 1`] = ` `;
